kernel/src/process_checking.rs

```diff
@@ -0,0 +1,360 @@
+use crate::process::{Process, State};
+use crate::utilities::cells::OptionalCell;
+use crate::utilities::cells::TakeCell;
+use crate::utilities::leasable_buffer::{LeasableBuffer, LeasableMutableBuffer};
+use crate::ErrorCode;
+use tock_tbf::types::TbfFooterV2Credentials;
+use tock_tbf::types::TbfFooterV2CredentialsType;
+
+/// What a AppCredentialsChecker decided a particular applications credential
+/// indicates about the runnability of an application binary.
+#[derive(Debug)]
+pub enum CheckResult {
+    /// Accept the credential and run the binary.
+    Accept,
+    /// Go to the next credential or in the case of the last one fall
+    /// back to the default policy.
+    Pass,
+    /// Reject the credential and do not run the binary.
+    Reject,
+}
+
+/// Receives callbacks on whether a credential was accepted or not.
+pub trait Client<'a> {
+    fn check_done(
+        &self,
+        result: Result<CheckResult, ErrorCode>,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    );
+}
+
+/// Implements a Credentials Checking Policy.
+pub trait AppCredentialsChecker<'a> {
+    fn set_client(&self, client: &'a dyn Client<'a>);
+    fn require_credentials(&self) -> bool;
+
+    fn check_credentials(
+        &self,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    ) -> Result<(), (ErrorCode, TbfFooterV2Credentials, &'a [u8])>;
+}
+
+/// Whether two processes have the same Application Identifier; two
+/// processes with the same Application Identifier cannot run concurrently.
+pub trait AppIdentification {
+    fn different_identifier(&self, process_a: &dyn Process, process_b: &dyn Process) -> bool;
+
+    // Return whether there is a currently running process that has
+    // the same application identifier as `process`. This means that
+    // if `process` is currently running, `has_unique_identifier`
+    // returns false.
+    fn has_unique_identifier(
+        &self,
+        process: &dyn Process,
+        processes: &[Option<&dyn Process>],
+    ) -> bool {
+        let len = processes.len();
+        if process.get_state() != State::Unstarted && process.get_state() != State::Terminated {
+            return false;
+        }
+
+        // Note that this causes `process` to compare against itself;
+        // however, since `process` should not be running, it will
+        // not check the identifiers and say they are different. This means
+        // this method returns false if the process is running.
+        for i in 0..len {
+            let checked_process = processes[i];
+            let diff = checked_process.map_or(true, |other| {
+                !other.is_running() || self.different_identifier(process, other)
+            });
+            if !diff {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+/// A compressed form of an Application Identifer.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct ShortID {
+    id: core::num::NonZeroU32,
+}
```

Comment:
I think this should be in process.rs so we can keep our various IDs together, and be able to describe how they differ and what each does.

Reply:
OK.

Reply:
Done.